### PR TITLE
DBG: Show the contents of references and pointers with LLDB

### LINKED ISSUE #9050: Automatically dereference pointer types when debugging with LLDB
## Environment

* **IntelliJ Rust plugin version:** 0.4.175.4733-221-nightly
* **Rust toolchain version:** 1.64.0-nightly (7b46aa594 2022-07-05) x86_64-apple-darwin
* **IDE name and version:** CLion 2022.1.3 (CL-221.5921.27)
* **Operating system:** macOS 11.4
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

Originally posted in https://youtrack.jetbrains.com/issue/CPP-29781

## Problem description

When debugging with LLDB, the contents of references and pointers are not rendered.

image

Note, when debugging with GDB instead of LLDB, the contents of references and pointers are rendered properly, except for the summary.

Screenshot 2023-01-30 at 14 58 49

## Steps to reproduce
```rust
fn print(s: &String, s2: String, u: u8) {
    println!("{}, {}, {}", s, s2, u);
}

pub fn main() {
    print(&"hello".to_string(), "hello".to_string(), 5);
}
```

### LINKED ISSUE #9252: Debugger does not show content of &&str
## Environment

* **IntelliJ Rust plugin version:** 0.4.176.4815-222
* **Rust toolchain version:** rustc 1.65.0-nightly (eaadb8947 2022-08-27)
* **IDE name and version:** CLion 2022.2.1 (Build #CL-222.3739.54, built on August 16, 2022)
* **Operating system:** ArchLinux kernel 5.19.4 64bit

## Problem description
Similar to https://github.com/intellij-rust/intellij-rust/issues/7139 , when using &&str it will not show the output.

For example, this will show:
```rust
let x = "hello";
```
image

But this won't:
```
let test = &"x";
```
image

## Steps to reproduce

1. Add `let test = &"x";`
2. Debug with that line

### PR DESCRIPTION
**Pull request by ortem**
Fixes #9050.
Partially fixes #9252 (LLDB only).

This PR brings automatic dereferencing of Rust references and pointers when debugging with LLDB. For example, now the debugger renders the contents of `&Vec`, `*const Vec` and `*mut Vec`, in addition to previously supported `Vec`.

Screenshot 2023-01-30 at 13 22 49

### Technical information

Previously, for each Rust type regex `$type` (for `String`, `Vec`, etc.), we ran the following LLDB commands:
```
(lldb) type summary add -F lldb_lookup.summary_lookup -e -x -h $type --category Rust
(lldb) type synthetic add -l lldb_lookup.synthetic_lookup -x $type --category Rust
```
They attach `summary_lookup` and `synthetic_lookup` Python functions as summary/synthetic providers for the given types. Actual providers are invoked by these lookup functions. This was done for two main reasons:
- for consistency, as GDB pretty-printing API requires us to use lookup functions (https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Pretty_002dPrinter.html);
- for flexibility, for example, to be able to choose the provider depending not only on the type regex, but also on its type class (`eTypeClassStruct`, `eTypeClassPointer`, etc.).

It turned out, however, that such flexibility is not needed for LLDB anymore. Moreover, it appeared that the current implementation breaks the automatic dereferencing of pointers provided by LLDB by default when registering pretty-printers just by regex.

So now, for each pair of `$type` and the corresponding `$provider`, we run these LLDB commands instead:
```
(lldb) type summary add -F lldb_formatters.lldb_providers.$provider -e -x -h $type --category Rust
(lldb) type synthetic add -l lldb_formatters.lldb_providers.$provider -x $type --category Rust
```

changelog: Show the contents of references and pointers when debugging with LLDB

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -337,6 +337,14 @@ project(":plugin") {
...
                 include("*.py")
             }
+            from("$rootDir/prettyPrinters/gdb_formatters") {
+                into("${pluginName.get()}/prettyPrinters/gdb_formatters")
+                include("*.py")
+            }
+            from("$rootDir/prettyPrinters/lldb_formatters") {
+                into("${pluginName.get()}/prettyPrinters/lldb_formatters")
+                include("*.py")
+            }
```
Comment (Undin) [inline build.gradle.kts:347]: It seems you can just replace initial `include("*.py")` with `include("**/*.py")` and avoid these new `from` blocks

**Review by Undin — COMMENTED** [review]
Could you resolve merge conflicts, please?

**Comment by ortem** [comment]
@Undin I've resolved merge conflicts and reordered the regexes for readability

File: prettyPrinters/lldb_formatters/__init__.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                                                                                                             regex)
+            )
+        if synth:
+            class_name = synth.__name__
+            debugger.HandleCommand(
+                'type synthetic add -l lldb_formatters.lldb_providers.{} -x "{}" --category Rust'.format(class_name,
+                                                                                                         regex)
+            )
+
+    register_providers(RustType.STR, StdStrSummaryProvider)
```
Comment (Undin) [inline prettyPrinters/lldb_formatters/__init__.py:27]: nit: maybe it's worth to pass providers with name argument syntax to increase readability?

**Review by Undin — APPROVED** [review]
Works for me!

Don't forget before merge:
- drop changes on CI
- describe that multiple dereferences don't work after these changes

File: pretty_printers_tests/tests/hash_map.rs [hunk 3]
```diff
@@ -3,8 +3,8 @@
 // lldb-command:run
 
 // lldb-command:print xs
-// lldbr-check:[...]xs = size=4 { [0] = { 0 = 1 1 = 10 } [1] = { 0 = 2 1 = 20 } [2] = { 0 = 3 1 = 30 } [3] = { 0 = 4 1 = 40 } }
-// lldbg-check:[...]$0 = size=4 { [0] = { 0 = 1 1 = 10 } [1] = { 0 = 2 1 = 20 } [2] = { 0 = 3 1 = 30 } [3] = { 0 = 4 1 = 40 } }
+// lldbr-check:[...]xs = size=4 { [0] = (0 = 1, 1 = 10) [1] = (0 = 2, 1 = 20) [2] = (0 = 3, 1 = 30) [3] = (0 = 4, 1 = 40) }
+// lldbg-check:[...]$0 = size=4 { [0] = (__0 = 1, __1 = 10) [1] = (__0 = 2, __1 = 20) [2] = (__0 = 3, __1 = 30) [3] = (__0 = 4, __1 = 40) }
```
[the remark this answers is not in the document]
Comment (ortem) [reply pretty_printers_tests/tests/hash_map.rs:7]: I should not affect users, so the UI representation for `HashMap` is supposed to be the same.

@intellij-rust/qa Please pay attention to this ^

File: pretty_printers_tests/tests/tuple.rs [hunk 4]
```diff
@@ -3,11 +3,12 @@
...
 // lldb-command:print t1
-// lldbr-check:[...]t1 = { 0 = { 0 = 1 1 = 2 } 1 = 3 }
-// lldbg-check:[...]$0 = { 0 = { 0 = 1 1 = 2 } 1 = 3 }
+// lldbr-check:[...]t1 = { 0 = (0 = 1, 1 = 2) 1 = 3 }
+// lldbg-check:[...]$0 = { __0 = (__0 = 1, __1 = 2) __1 = 3 }
+
 // lldb-command:print t2
 // lldbr-check:[...]t2 = { 0 = "abc" [...] 1 = 42 }
-// lldbg-check:[...]$1 = { 0 = "abc" [...] 1 = 42 }
+// lldbg-check:[...]$1 = { __0 = "abc" { data_ptr = 0x[...] "abc" length = 3 } __1 = 42 }
```
[the remark this answers is not in the document]
Comment (ortem) [reply pretty_printers_tests/tests/tuple.rs]: It seems redundant indeed, removed now

**Comment by ortem** [comment]
Note: this PR affects how the debugger renders "references to references" `&&T` and "pointers to pointers" and, more generally, multiple references and pointers. Consider this example:
```Rust
fn main() {
    let vec = vec![1,2,3];
    let vec_ref = &vec;
    let vec_ref_ref = &vec_ref;

    let s = String::from("abc");
    let s_ref = &s;
    let s_ref_ref = &s_ref;

    println!(); // break
}
```
| Before | After |
|--------|-------|
| Before      | After     |

Once this PR is merged, I will open an issue with a more detailed overview.

**Comment by ortem** [comment]
bors r=Undin

**Comment by mili-l** [comment]
https://github.com/rust-lang/rust/issues/106679

**Pull request merged**